proposals/25-ipld-in-ipfs.md

```diff
@@ -0,0 +1,148 @@
+Use of IPLD primarily flows through the [`DAGService`](https://github.com/ipfs/go-ipld-format/blob/master/merkledag.go), which sits on the BlockService, which is hardcoded to fetch through Bitswap, among other things.
+
+Several things sit on top of `DAGService` -- most prominently UnixFS.
+
+The first goal here is to get to a go-ipld-prime based `DAGService`, as close in interface to the existing DAGService, so that libraries dependent on this service would require only minimal changes.
+
+A first step is proposing an ipld-prime interface for this revamp’d DAG service, and to write a minimal scaffold translating from that interface to the current implementation.
+
+At that point, two people will continue working within the `dag service -> block service -> bitswap` and plumb through IPLD in that direction, while others will add more IPLD commands to the IPFS command line and handle propagating it to the various top level interfaces.
+
+We can also decide if we want to pursue updating some services that sit on top of `DAGService`.
+
+#### What does done look like?
+
+* P0: I can build IPFS without go-ipld-format
+* P1: I can import structured data (following links of generic codecs) directly from the IPFS command line
+* P1: structured data is importable and can be queried over the IPFS HTTP API 
+* P2: I can query structured data directly from the IPFS command line
+
+####  What does success look like?
+
+* People who have structured data or are looking at dynamic dApps do not leave the web3 ecosystem.
+  We see less new custom serialization code into IPFS files.
+* We see a growing percentage of gateway & DHT cid activity representing non-unixfs multicodec data.
+* We see less community frustration on inability to integrate DAG-JOSE
+* People who are considering dynamic dAPPs considers our stack
+* Filecoin is open to integrating go-ipld-prime in their codebase
+
+#### Counterpoints &amp; pre-mortem
+
+* People are struggling with static files, particularly in Filecoin, could leave before we can offer these services
+* We could find selectors or graphsync are not ultimately that useful a technology for working with structured data, or don’t unlock speed improvements
+* The impact of IPLD standardization will be felt strongest when we use go-ipld-prime across the stack.
+  We might finish our work in IPFS, but efforts to standardize in Filecoin could get delayed indefinitely by other project priorities.
+  This would leave standardization in a slightly better, but still incomplete state.
+* We may discover design issues in go-ipld-prime that require significant additional work to make it truly useful
+
+#### Alternatives
+
+* It’s possible we can just provide utilities in IPFS to import structured data and serialize to flat files?
+* It’s possible we can unlock some (thought not all) of the blocked projects without porting go-ipfs to go-ipld-prime
+
+#### Dependencies/prerequisites
+
+None, for the time being
```

Comment:
While not a hard blocker, should we use this as an opportunity to shore up ipld documentation?  I'm assuming lots of questions are going to come to @warpfork about ipld-prime.  Ideally all his answers should be responses back with links to public documentation (that was previously written or just written).  I thought I heard him say that there has been some IPLD documentation setup/structuring that has impaired this.  @warpfork : do you have any comments?